internlm/solver/optimizer/hybrid_zero_optim.py

```diff
@@ -385,35 +376,35 @@ def _reduce_grads_stored_in_bucket(self, reduce_rank=None, last_bucket=False):
         self._bucket_store.reset_by_rank(reduce_rank)
 
     def _reduce_grads_by_rank(self, reduce_rank, grads, bucket_size):
+        for bucket in self._bucket_in_progress:
+            bucket.commu_handle.wait()
+            bucket.unflatten_and_copy()
+            bucket.empty()
+        self._bucket_in_progress = []
+        self._param_store.clear_grads_of_previous_reduced_params()
+
         grad_buckets_by_dtype = split_half_float_double(grads)
 
         for tensor_list in grad_buckets_by_dtype:
             param_bucket = TensorBucket(size=bucket_size)
             for tensor in tensor_list:
                 param_bucket.add_to_bucket(tensor, allow_oversize=True)
-                if param_bucket.is_full_or_oversized():
-                    self._reduce_and_copy(bucket=param_bucket, reduce_rank=reduce_rank)
-                    param_bucket.empty()
             if not param_bucket.is_empty():
                 self._reduce_and_copy(bucket=param_bucket, reduce_rank=reduce_rank)
+            self._bucket_in_progress.append(param_bucket)
 
     def _reduce_and_copy(self, bucket: TensorBucket, reduce_rank):
-        if self._overlap_sync_grad:
-            self._comm_stream.synchronize()
-            self._param_store.clear_grads_of_previous_reduced_params()
```

Comment:
Notice `clear_grads_of_previous_reduced_params()` has been moved to step() function,
This can lead to de-synchronization between communications, making it impossible to release already reduced gradients during the backward. and it may impact the peak memory usage.